### PR TITLE
Adding put_record_batch API to firehose.rb

### PR DESCRIPTION
**Pull request by daynew**
Adding `put_record_batch` API to firehose.rb

* Adds support for batch requests to the AWS Firehose API.
* Adds support for using AWS Firehose streams other than analysis.events
* Adds unit tests for firehose.rb

## Context
I'm working on a different task to send batches of data to a new Firehose stream. I needed to update our existing Firehose client to support a batch API and also support a Firehose stream other than `analysis.evens`.

## Follow up work
* [Remove the DCDO experiment](https://codedotorg.atlassian.net/browse/FND-1190)

## Links
- [firehose put_record_batch docs](https://docs.aws.amazon.com/sdk-for-ruby/v3/api/Aws/Firehose/Client.html#put_record_batch-instance_method)

## Testing story
* Unit tests
* Manually tested that my new Firehose stream gets updates using this API.

# Reviewer Checklist:

- [ ] Tests provide adequate coverage
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: lib/cdo/firehose.rb [hunk 1]
```diff
@@ -18,31 +19,109 @@
...
+  # @param datas [array[hash]] The list of data to send to the stream.
+  def put_record_batch(stream_name, datas)
+    return if datas.nil? || datas.empty?
+    return unless Gatekeeper.allows('firehose', default: true)
+
+    # convert the given data into the format Firehose expects
+    records = []
+    datas.each do |data|
+      records << create_record_from_data(add_common_values(data))
+    end
```
Comment (wjordan) [inline lib/cdo/firehose.rb:66]: Use [`#map`](https://ruby-doc.org/core/Array.html#method-i-map):
```suggestion
    records = datas.map do |data|
      create_record_from_data(add_common_values(data))
    end
```

**Review by wjordan — COMMENTED** [review]
I have a general concern that the batching part of this PR duplicates part of the work found in #31958 (which was temporarily reverted due to a shutdown-related bug, and the only reason it hasn't been reopened is due to lower prioritization). That PR provided a general-purpose buffering + batching class that included implementations for Firehose `put_record_batch` as well as CloudWatch `put_metric_data`, and also supported batch-size limits (in bytes) and duration-based periodic flushing.

Unless there is a specific reason we think that earlier work should no longer be used, I'd much rather we revisit that earlier PR and diagnose+fix the bug than re-implement the same functionality from scratch.

**Comment by davidsbailey** [comment]
Hi @daynew , first question: what is the motivation for parameterizing the stream name?

**Comment by davidsbailey** [comment]
Second question: What is the idea behind directing calls to our `put_record` method to eventually call `@firehose.put_record_batch`? The docs at https://docs.aws.amazon.com/sdk-for-ruby/v2/api/Aws/Firehose/Client.html#put_record_batch-instance_method say:
>  To write single data records into a delivery stream, use [PutRecord](https://docs.aws.amazon.com/sdk-for-ruby/v2/api/Aws/Firehose/Client.html#put_record-instance_method).

If you want to gain an advantage here over our current implementation of `put_record`, I would imagine that you would need to store these individual requests in a buffer somewhere, and then send them to `put_record_batch` together once a certain amount of time or number of individual requests is reached. If we don't do that, then I'm not sure what we're gaining here.

**Pull request closed without merging**